### PR TITLE
add buffer pooling

### PR DESCRIPTION
**Pull request by niaow**
see #1

### CODE REVIEW

File: pool.go [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+
+	// recycle bufio reader and write buffer
+	iob.putBR.put(iob.br)
+	iob.putWBuf.put(iob.writeBuf)
+
+	// clear ioBuf to prevent reuse
+	*iob = ioBuf{}
+}
+
+// chanPool is a channel-based pool implementation for testing purposes
```
Comment (cristaloleg) [inline pool.go:149]: Is it used? I don't see any references in tests.

File: pool.go [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+
+import (
+	"bufio"
+	"errors"
+	"sync"
+)
+
+// Pool is a general-purpose interface for pooling.
+// Pool is implemented by sync.Pool.
+type Pool interface {
```
Comment (cristaloleg) [inline pool.go:11]: 'Cause this is an interface for a specific reason(byte pool), maybe we don't need to have params/results as an `interface{}` WDYT?

Comment (niaow) [reply pool.go:11]: 2 reasons:
1. It actually isnt - it is also used to pool *bufio.Reader
2. It is implemented by sync.Pool

File: pool.go (hunk 1 above)
Comment (niaow) [reply pool.go:149]: This PR is work in progress. Will be adding a test in a bit.

Comment (niaow) [reply pool.go:149]: Tests have been added.

**Pull request closed without merging**